File: compiler/nodes/expressions/when.go
package expressions

import (
	"wipple/codegen"
	"wipple/database"
	"wipple/nodes/patterns"
	"wipple/syntax"
	"wipple/typecheck"
	"wipple/visit"
)

type WhenExpressionNode struct {
	Input database.Node
	Arms  []Arm
	Facts *database.Facts

	inputTemporary database.Node
}

func (node *WhenExpressionNode) GetFacts() *database.Facts {
	return node.Facts
}

type Arm struct {
	Pattern database.Node
	Value   database.Node
	Span    database.Span
}

func ParseWhenExpression(parser *syntax.Parser) (*WhenExpressionNode, *syntax.Error) {
	span := parser.Spanned()

	_, err := parser.Token("WhenKeyword")
	if err != nil {
		return nil, err
	}

	parser.Commit("in this `when` expression")

	input, err := ParseAtomicExpression(parser)
	if err != nil {
		return nil, err
	}

	_, err = parser.Token("LeftBrace")
	if err != nil {
		return nil, err
	}

	arms, err := ParseArms(parser)
	if err != nil {
		return nil, err
	}

	_, err = parser.Token("RightBrace")
	if err != nil {
		return nil, err
	}

	return &WhenExpressionNode{
		Input: input,
		Arms:  arms,
		Facts: database.NewFacts(span()),
	}, nil
}

func ParseArm(parser *syntax.Parser) (Arm, *syntax.Error) {
	span := parser.Spanned()

	pattern, err := patterns.ParseAtomicPattern(parser)
	if err != nil {
		return Arm{}, err
	}

	parser.Commit("in this `when` arm")

	_, err = parser.Token("FunctionOperator")
	if err != nil {
		return Arm{}, err
	}

	parser.ConsumeLineBreaks()

	value, err := ParseExpression(parser)
	if err != nil {
		return Arm{}, err
	}

	return Arm{
		Pattern: pattern,
		Value:   value,
		Span:    span(),
	}, nil
}

func ParseArms(parser *syntax.Parser) ([]Arm, *syntax.Error) {
	return syntax.ParseLines(parser, 0, true, ParseArm)
}

func (node *WhenExpressionNode) Visit(visitor *visit.Visitor) {
	visitExpression(visitor, node)

	visitor.Visit(node.Input)
	visitor.Db.Graph.Edge(node.Input, node, "input")

	node.inputTemporary = &database.HiddenNode{
		Facts: database.NewFacts(database.GetSpanFact(node.Input)),
	}

	visitor.Db.Register(node.inputTemporary)

	visitor.Constraint(typecheck.NewGroupConstraint(node.inputTemporary, node.Input))

	visit.Matching(visitor, node.inputTemporary, true, func() struct{} {
		for _, arm := range node.Arms {
			visitor.PushScope()
			visitor.Visit(arm.Pattern)
			visitor.Db.Graph.Edge(arm.Pattern, node, "pattern")
			visitor.Visit(arm.Value)
			visitor.Db.Graph.Edge(arm.Value, node, "value")
			visitor.PopScope()

			visitor.Constraint(typecheck.NewGroupConstraint(arm.Value, node))
		}

		return struct{}{}
	})
}

func (node *WhenExpressionNode) Codegen(c *codegen.Codegen) error {
	if node.inputTemporary == nil {
		return c.Error(node)
	}

	span := database.GetSpanFact(node)

	c.WriteString(span, "await (async (")
	c.WriteNode(span, node.inputTemporary)
	c.WriteString(span, ") => {")
	c.WriteLine()

	armPatterns := make([]database.Node, 0, len(node.Arms))
	for _, arm := range node.Arms {
		armPatterns = append(armPatterns, arm.Pattern)
	}

	for _, temporary := range patterns.CollectTemporaries(armPatterns...) {
		if temporary == node.inputTemporary {
			continue
		}

		c.WriteString(span, "var ")
		c.WriteNode(span, temporary)
		c.WriteString(span, ";")
		c.WriteLine()
	}

	for _, arm := range node.Arms {
		c.WriteString(span, "if (true")
		if err := c.Write(arm.Pattern); err != nil {
			return err
		}
		c.WriteString(span, ") {")
		c.WriteLine()
		c.WriteString(span, "return ")
		if err := c.Write(arm.Value); err != nil {
			return err
		}
		c.WriteString(span, ";")
		c.WriteLine()
		c.WriteString(span, "}")
		c.WriteLine()
	}

	c.WriteString(span, `throw new Error("unreachable");`)
	c.WriteLine()

	c.WriteString(span, "})(")
	if err := c.Write(node.Input); err != nil {
		return err
	}
	c.WriteString(span, ")")

	return nil
}

File: compiler/src/nodes/expressions/when.rs
use crate::{
    codegen::{Codegen, CodegenCtx, CodegenResult, ir},
    database::{Node, NodeRef},
    nodes::{parse_atomic_expression, parse_atomic_pattern, parse_expression, visit_expression},
    syntax::{ParseError, Parser, TokenKind},
    typecheck::GroupConstraint,
    visit::{Visit, Visitor},
};

#[derive(Debug)]
pub struct WhenExpressionNode {
    pub input: NodeRef,
    pub arms: Vec<WhenArm>,
}

#[derive(Debug)]
pub struct WhenArm {
    pub pattern: NodeRef,
    pub value: NodeRef,
}

impl Node for WhenExpressionNode {}

pub fn parse_when_expression(parser: &mut Parser<'_>) -> Result<WhenExpressionNode, ParseError> {
    parser.token(TokenKind::WhenKeyword)?;
    parser.commit("in this `when` expression");

    let input = parse_atomic_expression(parser)?;

    parser.token(TokenKind::LeftBrace)?;
    let arms = parse_arms(parser)?;
    parser.token(TokenKind::RightBrace)?;

    Ok(WhenExpressionNode { input, arms })
}

pub fn parse_arm(parser: &mut Parser<'_>) -> Result<WhenArm, ParseError> {
    let pattern = parse_atomic_pattern(parser)?;

    parser.commit("in this `when` arm");
    parser.token(TokenKind::FunctionOperator)?;
    parser.consume_line_breaks();

    let value = parse_expression(parser)?;

    Ok(WhenArm { pattern, value })
}

pub fn parse_arms(parser: &mut Parser<'_>) -> Result<Vec<WhenArm>, ParseError> {
    parser.parse_lines(0, true, parse_arm)
}

impl Visit for WhenExpressionNode {
    fn visit(&self, node: &NodeRef, visitor: &mut Visitor<'_>) {
        visit_expression(node, visitor);

        visitor.visit(&self.input);
        visitor.edge(&self.input, node, "input");

        for arm in &self.arms {
            visitor.matching(&self.input, true, false, |visitor| {
                visitor.current_match().root = Some(self.input.clone());
                visitor.current_match().arm = Some(arm.pattern.clone());

                visitor.push_scope();
                visitor.visit(&arm.pattern);
                visitor.edge(&arm.pattern, node, "pattern");
                visitor.visit(&arm.value);
                visitor.edge(&arm.value, node, "value");
                visitor.pop_scope();

                visitor.constraint(GroupConstraint::new(arm.value.clone(), node.clone()));
            });
        }
    }
}

impl Codegen for WhenExpressionNode {
    fn codegen(&self, node: &NodeRef, ctx: &mut CodegenCtx<'_>) -> CodegenResult {
        ctx.codegen(&self.input)?;

        let branches = self
            .arms
            .iter()
            .map(|arm| {
                ctx.push_conditions();
                ctx.codegen(&arm.pattern)?;
                let conditions = ctx.pop_conditions();

                ctx.push_instructions();
                ctx.codegen(&arm.value)?;
                let instructions = ctx.pop_instructions();

                Ok((conditions, instructions, Some(arm.value.clone())))
            })
            .collect::<CodegenResult<Vec<_>>>()?;

        ctx.instruction(ir::Instruction::If {
            node: Some(node.clone()),
            branches,
            else_branch: None,
        });

        Ok(())
    }
}
